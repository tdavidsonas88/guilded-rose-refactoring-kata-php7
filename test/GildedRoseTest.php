<?php

namespace App;

use ApprovalTests\CombinationApprovals;
use \PHPUnit\Framework\TestCase;

/**
 * Class GildedRoseTest
 * @package App
 */
class GildedRoseTest extends TestCase {


    public function testUpdateQuality()
    {
        $names = [
            'foo',
            'Aged Brie',
            'Backstage passes',
            'Sulfuras',
            'Conjured'];
        $sellIns = [-1, 0, 5, 10, 11];
        $qualitys = [0, 1, 2, 6, 49, 50];

        CombinationApprovals::verifyAllCombinations3(function($a, $b, $c) {
            return $this->runUpdateQuality($a, $b, $c);
        }, $names, $sellIns, $qualitys);

    }


    private function runUpdateQuality(string $name, int $sellIn, int $quality)
    {
        /** @var Item[] $items */
        $items = [new Item($name, $sellIn, $quality)];
        $gildedRoseApp = new GildedRose($items);
        $gildedRoseApp->updateQuality();
        return $items[0]->__toString();
    }


    /**
     * Undocumented function
     *
     * @return void
     */
    public function getTestsData()
    {
        return [
            ['foo', 0, 5],
            ['abc', 0, 8],
        ];
    }

    // - Once the sell by date has passed, Quality degrades twice as fast
    /**
     * @dataProvider getTestsData
     * @param string $name
     * @param int $sellIn
     * @param int $quality
     */
    function testOnceTheSellByDateHasPassedQualityDegradesTwiceAsFast(string $name, int $sellIn, int $quality){
        $itemString = $this->runUpdateQuality($name, $sellIn, $quality);
        $itemValues = explode(',', $itemString);
        $this->assertSame($name.', '. strval($sellIn - 1) .', '. strval($quality -2), $itemString);
    }

    // At the end of each day our system lowers both values for every item
    function testAtTheEndOfEachDaySystemLowersValuesForEveryItem() {
        $itemString = $this->runUpdateQuality('foo', 2, 5);
        $this->assertSame('foo, 1, 4', $itemString);
    }
    // The Quality of an item is never negative
    function testTheQualityOfItemIsNeverNegative(){
        $items = array(new Item("foo", 2, 5));
        $gildedRose = new GildedRose($items);
        $gildedRose->updateQuality();
        $this->assertEquals(4, $items[0]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(3, $items[0]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(1, $items[0]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(0, $items[0]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(0, $items[0]->quality);
    }
    //- "Aged Brie" actually increases in Quality the older it gets
    function testAgeBrieQualityIncreasesTheOlderItGets() {
        $items = array(new Item("Aged Brie", 2, 5));
        $gildedRose = new GildedRose($items);
        $gildedRose->updateQuality();
        $this->assertEquals(6, $items[0]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(7, $items[0]->quality);
        //  'Aged Brie' increases by Two in quality after sell_in finished
        // nothing mentioned in the task but this is how it behaves
        $gildedRose->updateQuality();
        $this->assertEquals(9, $items[0]->quality);
    }
    // The Quality of an item is never more than 50
    function testQualityOfAnItemIsNeverMoreThan50() {
        $items = array(
            new Item("foo", 2, 50),
            new Item("Aged Brie", 2, 50),
            new Item("Sulfuras", 2, 80),
            new Item("Backstage passes", 2, 50),
            new Item("Conjured", 2, 50),
        );
        $gildedRose = new GildedRose($items);
        $gildedRose->updateQuality();
        $this->assertEquals(49, $items[0]->quality);
        $this->assertEquals(50, $items[1]->quality);
        $this->assertEquals(80, $items[2]->quality);
        $this->assertEquals(50, $items[3]->quality);
        $this->assertEquals(49, $items[4]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(50, $items[1]->quality);
        $this->assertEquals(80, $items[2]->quality);
        $this->assertEquals(50, $items[3]->quality);
        $this->assertEquals(48, $items[4]->quality);
    }
    // - "Sulfuras", being a legendary item, never has to be sold or decreases in Quality
    function testSulfurasNeverSoldOrDecreaseInQuality(){
        $items = array(
            new Item("Sulfuras", 2, 30)
        );
        $gildedRose = new GildedRose($items);
        $gildedRose->updateQuality();
        $this->assertEquals(80, $items[0]->quality);
        $this->assertEquals(2, $items[0]->sell_in);
        $gildedRose->updateQuality();
        $this->assertEquals(80, $items[0]->quality);
        $this->assertEquals(2, $items[0]->sell_in);
    }
    /**
     * - "Backstage passes", like aged brie, increases in Quality as its SellIn value approaches;
    Quality increases by 2 when there are 10 days or less and by 3 when there are 5 days or less but
    Quality drops to 0 after the concert
     */

    function testBackstagePasses(){
        $items = array(
            new Item("Backstage passes", 11, 30),
        );
        $gildedRose = new GildedRose($items);
        // 11
        $gildedRose->updateQuality();
        $this->assertEquals(10, $items[0]->sell_in);
        $this->assertEquals(31, $items[0]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(9, $items[0]->sell_in);
        $this->assertEquals(33, $items[0]->quality);

        $gildedRose->updateQuality();
        $this->assertEquals(8, $items[0]->sell_in);
        $this->assertEquals(35, $items[0]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(7, $items[0]->sell_in);
        $this->assertEquals(37, $items[0]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(6, $items[0]->sell_in);
        $this->assertEquals(39, $items[0]->quality);
        // 6
        $gildedRose->updateQuality();
        $this->assertEquals(5, $items[0]->sell_in);
        $this->assertEquals(41, $items[0]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(4, $items[0]->sell_in);
        $this->assertEquals(44, $items[0]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(3, $items[0]->sell_in);
        $this->assertEquals(47, $items[0]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(2, $items[0]->sell_in);
        $this->assertEquals(50, $items[0]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(1, $items[0]->sell_in);
        $this->assertEquals(50, $items[0]->quality);
        $gildedRose->updateQuality();
        $this->assertEquals(0, $items[0]->sell_in);
        $this->assertEquals(50, $items[0]->quality);
        // concert
        $gildedRose->updateQuality();
        $this->assertEquals(-1, $items[0]->sell_in);
        $this->assertEquals(0, $items[0]->quality);
    }
    // 	- "Conjured" items degrade in Quality twice as fast as normal items
    function testConjuredItemsDegradeInQualityTwiceAsFast(){
        $items = array(
            new Item("Conjured", 2, 30)
        );
        $gildedRose = new GildedRose($items);
        $gildedRose->updateQuality();
        $this->assertEquals(28, $items[0]->quality);

        $gildedRose->updateQuality();
        $this->assertEquals(26, $items[0]->quality);
    }
}
