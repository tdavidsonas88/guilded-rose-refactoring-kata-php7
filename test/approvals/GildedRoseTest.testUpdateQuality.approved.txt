[foo, -1, 0] => foo, -2, 0
[foo, -1, 1] => foo, -2, 0
[foo, -1, 2] => foo, -2, 0
[foo, -1, 6] => foo, -2, 4
[foo, -1, 49] => foo, -2, 47
[foo, -1, 50] => foo, -2, 48
[foo, 0, 0] => foo, -1, 0
[foo, 0, 1] => foo, -1, 0
[foo, 0, 2] => foo, -1, 0
[foo, 0, 6] => foo, -1, 4
[foo, 0, 49] => foo, -1, 47
[foo, 0, 50] => foo, -1, 48
[foo, 5, 0] => foo, 4, 0
[foo, 5, 1] => foo, 4, 0
[foo, 5, 2] => foo, 4, 1
[foo, 5, 6] => foo, 4, 5
[foo, 5, 49] => foo, 4, 48
[foo, 5, 50] => foo, 4, 49
[foo, 10, 0] => foo, 9, 0
[foo, 10, 1] => foo, 9, 0
[foo, 10, 2] => foo, 9, 1
[foo, 10, 6] => foo, 9, 5
[foo, 10, 49] => foo, 9, 48
[foo, 10, 50] => foo, 9, 49
[foo, 11, 0] => foo, 10, 0
[foo, 11, 1] => foo, 10, 0
[foo, 11, 2] => foo, 10, 1
[foo, 11, 6] => foo, 10, 5
[foo, 11, 49] => foo, 10, 48
[foo, 11, 50] => foo, 10, 49
[Aged Brie, -1, 0] => Aged Brie, -2, 2
[Aged Brie, -1, 1] => Aged Brie, -2, 3
[Aged Brie, -1, 2] => Aged Brie, -2, 4
[Aged Brie, -1, 6] => Aged Brie, -2, 8
[Aged Brie, -1, 49] => Aged Brie, -2, 50
[Aged Brie, -1, 50] => Aged Brie, -2, 50
[Aged Brie, 0, 0] => Aged Brie, -1, 2
[Aged Brie, 0, 1] => Aged Brie, -1, 3
[Aged Brie, 0, 2] => Aged Brie, -1, 4
[Aged Brie, 0, 6] => Aged Brie, -1, 8
[Aged Brie, 0, 49] => Aged Brie, -1, 50
[Aged Brie, 0, 50] => Aged Brie, -1, 50
[Aged Brie, 5, 0] => Aged Brie, 4, 1
[Aged Brie, 5, 1] => Aged Brie, 4, 2
[Aged Brie, 5, 2] => Aged Brie, 4, 3
[Aged Brie, 5, 6] => Aged Brie, 4, 7
[Aged Brie, 5, 49] => Aged Brie, 4, 50
[Aged Brie, 5, 50] => Aged Brie, 4, 50
[Aged Brie, 10, 0] => Aged Brie, 9, 1
[Aged Brie, 10, 1] => Aged Brie, 9, 2
[Aged Brie, 10, 2] => Aged Brie, 9, 3
[Aged Brie, 10, 6] => Aged Brie, 9, 7
[Aged Brie, 10, 49] => Aged Brie, 9, 50
[Aged Brie, 10, 50] => Aged Brie, 9, 50
[Aged Brie, 11, 0] => Aged Brie, 10, 1
[Aged Brie, 11, 1] => Aged Brie, 10, 2
[Aged Brie, 11, 2] => Aged Brie, 10, 3
[Aged Brie, 11, 6] => Aged Brie, 10, 7
[Aged Brie, 11, 49] => Aged Brie, 10, 50
[Aged Brie, 11, 50] => Aged Brie, 10, 50
[Backstage passes, -1, 0] => Backstage passes, -2, 0
[Backstage passes, -1, 1] => Backstage passes, -2, 0
[Backstage passes, -1, 2] => Backstage passes, -2, 0
[Backstage passes, -1, 6] => Backstage passes, -2, 0
[Backstage passes, -1, 49] => Backstage passes, -2, 0
[Backstage passes, -1, 50] => Backstage passes, -2, 0
[Backstage passes, 0, 0] => Backstage passes, -1, 0
[Backstage passes, 0, 1] => Backstage passes, -1, 0
[Backstage passes, 0, 2] => Backstage passes, -1, 0
[Backstage passes, 0, 6] => Backstage passes, -1, 0
[Backstage passes, 0, 49] => Backstage passes, -1, 0
[Backstage passes, 0, 50] => Backstage passes, -1, 0
[Backstage passes, 5, 0] => Backstage passes, 4, 3
[Backstage passes, 5, 1] => Backstage passes, 4, 4
[Backstage passes, 5, 2] => Backstage passes, 4, 5
[Backstage passes, 5, 6] => Backstage passes, 4, 9
[Backstage passes, 5, 49] => Backstage passes, 4, 52
[Backstage passes, 5, 50] => Backstage passes, 4, 50
[Backstage passes, 10, 0] => Backstage passes, 9, 2
[Backstage passes, 10, 1] => Backstage passes, 9, 3
[Backstage passes, 10, 2] => Backstage passes, 9, 4
[Backstage passes, 10, 6] => Backstage passes, 9, 8
[Backstage passes, 10, 49] => Backstage passes, 9, 51
[Backstage passes, 10, 50] => Backstage passes, 9, 50
[Backstage passes, 11, 0] => Backstage passes, 10, 1
[Backstage passes, 11, 1] => Backstage passes, 10, 2
[Backstage passes, 11, 2] => Backstage passes, 10, 3
[Backstage passes, 11, 6] => Backstage passes, 10, 7
[Backstage passes, 11, 49] => Backstage passes, 10, 50
[Backstage passes, 11, 50] => Backstage passes, 10, 50
[Sulfuras, -1, 0] => Sulfuras, -1, 80
[Sulfuras, -1, 1] => Sulfuras, -1, 80
[Sulfuras, -1, 2] => Sulfuras, -1, 80
[Sulfuras, -1, 6] => Sulfuras, -1, 80
[Sulfuras, -1, 49] => Sulfuras, -1, 80
[Sulfuras, -1, 50] => Sulfuras, -1, 80
[Sulfuras, 0, 0] => Sulfuras, 0, 80
[Sulfuras, 0, 1] => Sulfuras, 0, 80
[Sulfuras, 0, 2] => Sulfuras, 0, 80
[Sulfuras, 0, 6] => Sulfuras, 0, 80
[Sulfuras, 0, 49] => Sulfuras, 0, 80
[Sulfuras, 0, 50] => Sulfuras, 0, 80
[Sulfuras, 5, 0] => Sulfuras, 5, 80
[Sulfuras, 5, 1] => Sulfuras, 5, 80
[Sulfuras, 5, 2] => Sulfuras, 5, 80
[Sulfuras, 5, 6] => Sulfuras, 5, 80
[Sulfuras, 5, 49] => Sulfuras, 5, 80
[Sulfuras, 5, 50] => Sulfuras, 5, 80
[Sulfuras, 10, 0] => Sulfuras, 10, 80
[Sulfuras, 10, 1] => Sulfuras, 10, 80
[Sulfuras, 10, 2] => Sulfuras, 10, 80
[Sulfuras, 10, 6] => Sulfuras, 10, 80
[Sulfuras, 10, 49] => Sulfuras, 10, 80
[Sulfuras, 10, 50] => Sulfuras, 10, 80
[Sulfuras, 11, 0] => Sulfuras, 11, 80
[Sulfuras, 11, 1] => Sulfuras, 11, 80
[Sulfuras, 11, 2] => Sulfuras, 11, 80
[Sulfuras, 11, 6] => Sulfuras, 11, 80
[Sulfuras, 11, 49] => Sulfuras, 11, 80
[Sulfuras, 11, 50] => Sulfuras, 11, 80
[Conjured, -1, 0] => Conjured, -2, 0
[Conjured, -1, 1] => Conjured, -2, 0
[Conjured, -1, 2] => Conjured, -2, 0
[Conjured, -1, 6] => Conjured, -2, 4
[Conjured, -1, 49] => Conjured, -2, 47
[Conjured, -1, 50] => Conjured, -2, 48
[Conjured, 0, 0] => Conjured, -1, 0
[Conjured, 0, 1] => Conjured, -1, 0
[Conjured, 0, 2] => Conjured, -1, 0
[Conjured, 0, 6] => Conjured, -1, 2
[Conjured, 0, 49] => Conjured, -1, 45
[Conjured, 0, 50] => Conjured, -1, 46
[Conjured, 5, 0] => Conjured, 4, 0
[Conjured, 5, 1] => Conjured, 4, 0
[Conjured, 5, 2] => Conjured, 4, 0
[Conjured, 5, 6] => Conjured, 4, 4
[Conjured, 5, 49] => Conjured, 4, 47
[Conjured, 5, 50] => Conjured, 4, 48
[Conjured, 10, 0] => Conjured, 9, 0
[Conjured, 10, 1] => Conjured, 9, 0
[Conjured, 10, 2] => Conjured, 9, 0
[Conjured, 10, 6] => Conjured, 9, 4
[Conjured, 10, 49] => Conjured, 9, 4
[Conjured, 10, 50] => Conjured, 9, 48
[Conjured, 11, 0] => Conjured, 10, 0
[Conjured, 11, 1] => Conjured, 10, 0
[Conjured, 11, 2] => Conjured, 10, 0
[Conjured, 11, 6] => Conjured, 10, 3
[Conjured, 11, 49] => Conjured, 10, 47
[Conjured, 11, 50] => Conjured, 10, 48
